File: docs-examples/example-java/src/test/java/io/micronaut/configuration/rabbitmq/docs/acknowledge/publisher/PublisherAcknowledgeSpec.java
package io.micronaut.configuration.rabbitmq.docs.acknowledge.publisher;

import io.micronaut.configuration.rabbitmq.AbstractRabbitMQTest;
import io.micronaut.configuration.rabbitmq.annotation.Binding;
import io.micronaut.context.ApplicationContext;
import io.reactivex.Completable;
import io.reactivex.CompletableObserver;
import io.reactivex.Maybe;
import io.reactivex.MaybeObserver;
import io.reactivex.disposables.Disposable;
import org.junit.jupiter.api.Test;
import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;
import reactor.core.publisher.Mono;

import java.util.concurrent.atomic.AtomicInteger;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

public class PublisherAcknowledgeSpec extends AbstractRabbitMQTest {

    @Test
    void testPublisherAcknowledgement() {
        ApplicationContext applicationContext = startContext();
        AtomicInteger successCount = new AtomicInteger(0);
        AtomicInteger errorCount = new AtomicInteger(0);

// tag::producer[]
ProductClient productClient = applicationContext.getBean(ProductClient.class);
Completable completable = productClient.send("completable body".getBytes());
Maybe<Void> maybe = productClient.sendMaybe("maybe body".getBytes());
Mono<Void> mono = productClient.sendMono("mono body".getBytes());
Publisher<Void> publisher = productClient.sendPublisher("publisher body".getBytes());

completable.subscribe(new CompletableObserver() {
    @Override
    public void onSubscribe(Disposable d) { }

    @Override
    public void onComplete() {
        // if the publish was acknowledged
        successCount.incrementAndGet();
    }

    @Override
    public void onError(Throwable e) {
        // if an error occurs
        errorCount.incrementAndGet();
    }
});
maybe.subscribe(new MaybeObserver<Void>() {
    @Override
    public void onSubscribe(Disposable d) { }

    @Override
    public void onSuccess(Void aVoid) {
        throw new UnsupportedOperationException("Should never be called");
    }

    @Override
    public void onError(Throwable e) {
        // if an error occurs
        errorCount.incrementAndGet();
    }

    @Override
    public void onComplete() {
        // if the publish was acknowledged
        successCount.incrementAndGet();
    }
});
Subscriber<Void> subscriber = new Subscriber<Void>() {
    @Override
    public void onSubscribe(Subscription subscription) { }

    @Override
    public void onNext(Void aVoid) {
        throw new UnsupportedOperationException("Should never be called");
    }

    @Override
    public void onError(Throwable throwable) {
        // if an error occurs
        errorCount.incrementAndGet();
    }

    @Override
    public void onComplete() {
        // if the publish was acknowledged
        successCount.incrementAndGet();
    }
};
mono.subscribe(subscriber);
publisher.subscribe(subscriber);
// end::producer[]

        try {
            await().atMost(5, SECONDS).until(() ->
                    errorCount.get() == 0 &&
                    successCount.get() == 4
            );
        } finally {
            applicationContext.close();
        }
    }

}

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/publisher/acknowledge/PublisherAcknowledgeSpec.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.rabbitmq.docs.Await import Await
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test
from org.reactivestreams import Subscriber, Subscription

from .ProductClient import ProductClient


@MicronautTest(environments=["rabbitmq"])
@Property(name="spec.name", value="PublisherAcknowledgeSpec")
class PublisherAcknowledgeSpec:
    product_client: Annotated[ProductClient, Inject]

    @Test
    def test_publisher_acknowledgement(self) -> None:
        counts = {"success": 0, "error": 0}

# tag::producer[]
        publisher = self.product_client.send_publisher(b"publisher body")
        future = self.product_client.send_future(b"future body")

        class AcknowledgementSubscriber(Subscriber):

            def onSubscribe(self, subscription: Subscription) -> None:
                pass

            def onNext(self, item: object) -> None:
                raise RuntimeError("Should never be called")

            def onError(self, throwable: Exception) -> None:
                # if an error occurs
                counts["error"] += 1

            def onComplete(self) -> None:
                # if the publish was acknowledged
                counts["success"] += 1

        publisher.subscribe(AcknowledgementSubscriber())

        def completed(value, throwable) -> None:
            if throwable is None:
                counts["success"] += 1
            else:
                counts["error"] += 1

        future.whenComplete(completed)
# end::producer[]

        Await.until(lambda: counts["error"] == 0 and counts["success"] == 2)
